cmd/tsbs_load_mongo/creator.go

```diff
@@ -36,74 +53,68 @@ func (d *dbCreator) DBExists(dbName string) bool {
 		return err
 	}
-	for _, name := range collections {
-		d.session.DB(dbName).C(name).DropCollection()
+	for _, coll := range collection_names {
+		log.Printf("collection found :  %s", coll)
+		log.Println("deleting the previous collection")
+		err := d.client.Database(dbName).Collection(coll).Drop(d.ctx)
+		if err != nil {
+			log.Printf("Could not delete collection : %s", err.Error())
+		}
 	}
-
 	return nil
 }
 
 func (d *dbCreator) CreateDB(dbName string) error {
-	cmd := make(bson.D, 0, 4)
-	cmd = append(cmd, bson.DocElem{Name: "create", Value: collectionName})
-
-	// wiredtiger settings
-	cmd = append(cmd, bson.DocElem{
-		Name: "storageEngine", Value: map[string]interface{}{
-			"wiredTiger": map[string]interface{}{
-				"configString": "block_compressor=snappy",
-			},
-		},
-	})
-
-	err := d.session.DB(dbName).Run(cmd, nil)
+	//log.Println("tsbs_load_mongo/creator/CreateDB")
+	//Starting in MongoDB 3.2, the WiredTiger storage engine is the default storage engine
+	err := d.client.Database(dbName).CreateCollection(d.ctx, collectionName)
 	if err != nil {
 		if strings.Contains(err.Error(), "already exists") {
+			log.Printf("collection %s already exists", dbName)
 			return nil
 		}
+		log.Printf("create collection err: %v", err)
 		return fmt.Errorf("create collection err: %v", err)
 	}
-
-	collection := d.session.DB(dbName).C(collectionName)
-	var key []string
+	collection := d.client.Database(dbName).Collection(collectionName)
+	var key bson.D
 	if documentPer {
-		key = []string{"measurement", "tags.hostname", timestampField}
+		key = bson.D{{"measurement", 1}, {"tags.hostname", 1}, {timestampField, 1}}
 	} else {
-		key = []string{aggKeyID, "measurement", "tags.hostname"}
+		key = bson.D{{aggKeyID, 1}, {"measurement", 1}, {"tags.hostname", 1}}
 	}
-
-	index := mgo.Index{
-		Key:        key,
-		Unique:     false, // Unique does not work on the entire array of tags!
-		Background: false,
-		Sparse:     false,
+	index := mongo.IndexModel{
+		Keys:    key,
+		Options: options.Index().SetName("default_index"),
 	}
-	err = collection.EnsureIndex(index)
+	idxview := collection.Indexes()
+	_, err = idxview.CreateOne(d.ctx, index)
 	if err != nil {
-		return fmt.Errorf("create basic index err: %v", err)
+		log.Printf("create index err: %v", err)
+		panic(err)
 	}
-
 	// To make updates for new records more efficient, we need a efficient doc
 	// lookup index
 	if !documentPer {
-		err = collection.EnsureIndex(mgo.Index{
-			Key:        []string{aggDocID},
-			Unique:     false,
-			Background: false,
-			Sparse:     false,
+		_, err := idxview.CreateOne(d.ctx, mongo.IndexModel{
+			Keys:    bson.D{{aggDocID, 1}},
+			Options: options.Index().SetName("doc_lookup_index"),
 		})
 		if err != nil {
-			return fmt.Errorf("create agg doc index err: %v", err)
+			log.Printf("create index err: %v", err)
+			panic(err)
 		}
 	}
-
 	return nil
 }
 
 func (d *dbCreator) Close() {
-	d.session.Close()
+	//closing the database connection here
```

Comment:
Can we fix the error and wait for the bulk finish and then close the connection? or force close in case that is what the user wants to do?

Reply:
This is tricky, I'm going to look into it when I have more time